### PR TITLE
Generate a summary of built-in hints

### PR DESCRIPTION
**Pull request by zliu41**
Based on suggestions from @shayne-fletcher and @ndmitchell in #1037.

- Generated from `cabal run hlint test -- --generate-summary`
- One entry per (hint name, severity, does it support refactoring)
- No change to CI yet (i.e., it currently does not check that the summary is up to date). The check can be done either in `cabal run hlint test` or in the CI script. Not sure which one is preferred. Either way it can be done separately.

### CODE REVIEW

File: src/Test/Util.hs [hunk 1]
```diff
@@ -1,22 +1,41 @@
-{-# LANGUAGE RecordWildCards, GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE RecordWildCards, GeneralizedNewtypeDeriving, StrictData #-}
```
Comment (ndmitchell) [inline src/Test/Util.hs:1]: Can we just put ! on individual fields. I've been burnt by StrictData in the past.

File: src/Hint/Pragma.hs [hunk 2]
```diff
@@ -15,7 +15,7 @@
 {-# OPTIONS_YHC -cpp #-}
 {-# OPTIONS_GHC -XFoo #-} -- {-# LANGUAGE Foo #-}
 {-# OPTIONS_GHC -fglasgow-exts #-} -- ??? @NoRefactor
-{-# LANGUAGE RebindableSyntax, EmptyCase, DuplicateRecordFields, RebindableSyntax #-} -- {-# LANGUAGE RebindableSyntax, EmptyCase, DuplicateRecordFields #-}
```
Comment (ndmitchell) [inline src/Hint/Pragma.hs:18]: Why this change?

**Comment by ndmitchell** [comment]
Cool! One little remark, but lets merge and iterate - this is a great start!

**Comment by shayne-fletcher** [comment]
I agree, this is a great start! Couple of thoughts for the next iterations:
  - should `--generate-summary` take a path argument rather than hardcoding "builtin.md"?
  - can the summary be generalized someway, somehow to consider *all* hints in scope (including those that are not builtin)?

**Comment by shayne-fletcher** [comment]
Guess Appveyor needs a retry.
Screen Shot 2020-06-23 at 10 23 08 AM

File: src/Hint/Pragma.hs (hunk 2 above)
Comment (zliu41) [reply src/Hint/Pragma.hs:18]: The shorter version fits in a single line in the markdown table. I added the long version back below it.

File: src/Test/Util.hs (hunk 1 above)
Comment (zliu41) [reply src/Test/Util.hs]: Done!

**Comment by zliu41** [comment]
> can the summary be generalized someway, somehow to consider all hints in scope (including those that are not builtin)?

I think this is easy to do; I didn't include lhs => rhs hints because `hlint.yaml` already lists those hints.

**Comment by shayne-fletcher** [comment]
> > can the summary be generalized someway, somehow to consider all hints in scope (including those that are not builtin)?
> 
> I think this is easy to do; I didn't include lhs => rhs hints because `hlint.yaml` already lists those hints.

Personally, I think "in for a penny, in for a pound". A complete report including those that aren't builtin on one handy web-page with a listing of their status (and perhaps indicators of whether they are builtin or custom) would be happy days for me 😁 . @ndmitchell thoughts?

**Comment by ndmitchell** [comment]
I'd like to do all hints, builtin and otherwise, with examples for all. The lhs/rhs ones are pretty easy to do builtin ones, because the LHS is a good example. But that can come later.

Unfortunately Windows and Haskell is now a bit of a dead zone. I happily merge with Appveyor failures, but feel bad about it.

**Comment by zliu41** [comment]
Sounds good, let's add the lhs/rhs hints in the near future.

**Comment by ndmitchell** [comment]
Thanks!
